### PR TITLE
Added Session object for cookie storage and defaults

### PR DESCRIPTION
**Pull request by Drizzt1991**
Simple snippet of code:

    >>> import asyncio
    >>> from aiohttp import client
    >>> loop = asyncio.get_event_loop()
    >>> s = client.Session(loop=loop)
    >>> r = loop.run_until_complete(s.get("http://google.com"))
    >>> s.cookies == r.cookies
    True

### CODE REVIEW

File: aiohttp/client.py [hunk 1]
```diff
@@ -83,69 +83,227 @@ def request(method, url, *,
...
+                                       allow_redirects=allow_redirects,
+                                       **kwargs)
+        return resp
 
-            url = urllib.parse.urldefrag(r_url)[0]
-            if url:
-                yield from asyncio.async(resp.release(), loop=loop)
-                continue
+    @asyncio.coroutine
+    def options(self, url, allow_redirects=True, **kwargs):
```
Comment (kxepal) [inline aiohttp/client.py]: Why allow_redirects is positional argument here when data for post/put isn't?

**Comment by kxepal** [comment]
Question: if I'm using Cookie based auth and my cookie eventually expired, how could I renew it? Same issue with JWT tokens and other things with TTL.

The solution I see is to provide a callback to session and let it call it in case of first 401 error to reauth the session.

**Comment by asvetlov** [comment]
I think session expiration is too complex in general for implementing it into client itself.
You may wrap your calls into try/except block for renewing session by default.
Or derive from Session and override `.request()` method.

**Comment by kxepal** [comment]
@asvetlov if it was so easy: I can't wrap every call with try/except/renew since I actually don't know when and where session will get expired. That's why passing some callback coroutine to trigger on auth error could be helpful.

Yes, overriding .request() is what I'm doing now (: But thought that's quite common problem to have some base solution on the ground.

File: aiohttp/client.py [hunk 2]
```diff
@@ -83,69 +83,227 @@ def request(method, url, *,
...
+        resp = yield from self.request('OPTIONS', url,
+                                       allow_redirects=allow_redirects,
+                                       **kwargs)
+        return resp
 
-        break
+    @asyncio.coroutine
+    def head(self, url, allow_redirects=False, **kwargs):
+        resp = yield from self.request('HEAD', url,
```
Comment (fafhrd91) [inline aiohttp/client.py]: use hdrs.METH_XXXX instead of strings

**Comment by asvetlov** [comment]
Actually you have a need for two callbacks: 
- one for detection for session expiring. In general techically it may be not only 401 response code but everything including some specific body for 200 OK. Site builders sometimes don't follow any standards.
- another one for session renewal.

I think it's overengeneering. At least we can add those callbacks later.

**Comment by kxepal** [comment]
Eh, forgot about bad sites. Ok, may be I just want too much now (: Anyway, Session is a good step forward.

**Comment by fafhrd91** [comment]
i'd like to use something like base_url

```
session = Session('http://python.com/')
session.get('/docs/index.html')
session.post('/login.html')
```

**Comment by kxepal** [comment]
@fafhrd91 I'll a bit extend your proposal: how about something like [this](https://github.com/aio-libs/aiocouchdb/blob/master/aiocouchdb/client.py#L163)? Two points:
1. Construction urls with `__call__` using unescaped path segments
2. Extraction credentials from URL into Authorization header

However, all these features in additional to http-method-like class methods are a functionality of HTTP resources, not a session which should only keep and track a state, ~~made requests~~ apply it onto outgoing requests and update from incoming responses.

**Comment by asvetlov** [comment]
I'm strongly against base_url in Session class, it should be added in descendant.
Session is for processing cookies/keepalives only.

Adding base url is a can of worms: you have to process:
- absolute urls in .get() etc
- schema changing (from http to https for example)
- another question is: should base_url be changed on redirect to outside site?

I feel adding base_url looks very appealing but think that should be implemented in another class. Processing different schema (http, https, //) and working with absolute and relative urls is not trivial.

**Comment by fafhrd91** [comment]
how often you change schema? maybe if you building proxy. if you build client for existing rest-style service it is much easier to use base_url. my feeling people use it as client for existing service more often that anything else.

**Comment by asvetlov** [comment]
My guess is to add base_url in new class derived from Session

On Tue, Apr 14, 2015 at 5:47 PM, Nikolay Kim notifications@github.com
wrote:

> how often you change schema? maybe if you building proxy. if you build
> client for existing rest-style service it is much easier to use base_url.
> my feeling people use it as client for existing service more often that
> anything else.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/KeepSafe/aiohttp/pull/328#issuecomment-93077818.

## 

Thanks,
Andrew Svetlov

**Comment by fafhrd91** [comment]
ok. let just rename to ClientSession

**Comment by asvetlov** [comment]
ok with ClientSession name.

**Comment by fafhrd91** [comment]
i'm planing to release next 0.15 release during this week, let's include session as well.

**Comment by tvoinarovskyi** [comment]
Fixed minor issues, pointed out before.
Added docs on ClientSessions.

**Pull request merged**